### ISSUE TITLE
Partial goal

### ISSUE THREAD

**Issue by guestisp** [moosefs/moosefs#116]
Let's assume a goal of 4
Is possibile to set how many chuncks should be wrote before returning an ACK to the client?

In example, to improve performance, we can set a redundancy level of 2, in this case, after 2 succsefully wrote chunks, the ACK is returned to the client while the remaining 2 chunks are still being written in a sort of writeback

(Lizard has this and i think something similiar is also supported by ceph)

**Comment by oszafraniec** [comment]
I think you can achieve this by setting up classes, not goals. You have create/keep/archive states that can have different goals and also use different servers according to labels assigned to those states.

So you can "create" a file with goal 2 (so client waits only for 2 copies to be created), "keep" it with goal 4 (extra copies will be made after some time) and after some longer time f.ex. "archive" it as goal 3 to save space on "old data".

**Comment by guestisp** [comment]
This is absolutely awesome.
The "after some time" for "keep" goal is customizable, right ?

**Comment by oszafraniec** [comment]
No, only archive time can be set up. Keep is made "asap" i think.

**Comment by oxide94** [comment]
Hi @guestisp,

As @oszafraniec wrote - keep is made asap, archive can be done after certain time (in MooseFS 3.x only basing on `mtime`). In MooseFS 4.x we extended it and it can be based on `ctime` (change time), `atime` (access time) and `mtime` (modification time). In MooseFS 4.x it can also be one-way or reversible).

Best,
Peter / MooseFS Team

**Comment by oxide94** [comment]
BTW - as far as I remember there is no difference in speed between writing to file in goal 2 and to file in goal 4. I'm not 100% sure here (we ran such tests quite long time ago), but I definitely remember, that there is no difference in speed between writing to file in goal 1 vs. to file in goal 2.

Peter

**Comment by guestisp** [comment]
There is a speed difference if one of the affected chunkserver is on a slow connection. In that case, everything will slow down, because you have to wait that all chunks are properly stored.

When using an high level of reduncancy, you can safely assume that data is secured even after only 2 copies are made (2 is better than 1), then the third or fourth could also be made in a "non-blocking" way.

I hope is clear what i'm trying to describe.

Replicating "keep" goals ASAP is perfect, as long the replication doesn't "block" the client.

How "keep" copies are managed by MooseFS ? In case of goal=2, keep=4, we should end with 4 replica of the same file. When a "keep" copy is down, MooseFS thread that as an undergoal ?

**Comment by oszafraniec** [comment]
Yes, undergoal. And I think I know what you want to do... Let's say you have 2 "local" CHSRV and 2 "remote/slower" CHSRV in different DC. Band to other DC is not as quick as local. You want to create chunks locally and then do remove extra goal copies in keep. That's what you will get via classes and labels.

But remember that reading data it can happen that the client will read them from remote/slower CHSRV. There is no way to tell MFS to read first from local CHSRV and in case of undergoal to go to remote/slower CHSRV.

Dont know if this is your scenario, but keep in mind reading and updateing the file after it is made to "keep" state.

**Comment by guestisp** [comment]
> Yes, undergoal. And I think I know what you want to do... Let's say you have 2 "local" CHSRV and 2 "remote/slower" CHSRV in different DC. Band to other DC is not as quick as local. You want to create chunks locally and then do remote extra goal copies in keep. That's what you will get via classes and labels.

Mine was just an example but yes, a geo replication across two DC for distaster recovery is something i'll use in the future.

> There is no way to tell MFS to read first from local CHSRV and in case of undergoal to go to remote/slower CHSRV.

Is something you are planning to add?

> Dont know if this is your scenario, but keep in mind reading and updateing the file after it is made to "keep" state.

Can you elaborate more  ? Is not really clear what you mean.

**Comment by oxide94** [comment]
> There is no way to tell MFS to read first from local CHSRV and in case of undergoal to go to remote/slower CHSRV.

There is a way to prefer certain Chunkserver types (i.e. labeled as LABELEXPR) during I/O.

From `man mfsmount`:
```
       -o mfspreflabels=LABELEXPR
              specify preferred labels for choosing chunkservers during I/O
```

There is also a whole paragraph on it in [this manual](https://moosefs.com/Content/Downloads/moosefs-3-0-users-manual.pdf): **4.2.10** "Preferred labels during read/write (in `mfsmount`)".

Best regards,
Peter / MooseFS Team

**Comment by oszafraniec** [comment]
Didn't know about that :) Thank you AGAIN @oxide94  ;)
